File: app/src/main/java/me/bkkn/ui/MainActivity.java
package me.bkkn.ui;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.SearchView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.menu.MenuItemWrapperICS;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;

import com.google.android.material.snackbar.Snackbar;

import me.bkkn.App;
import me.bkkn.R;
import me.bkkn.domain.entity.Note;
import me.bkkn.ui.details.NoteDetailsFragment;
import me.bkkn.ui.dialog.AlertDialogFragment;
import me.bkkn.ui.list.NotesFragment;

public class MainActivity
        extends AppCompatActivity
        implements NotesFragment.Controller, NoteDetailsFragment.Controller, AlertDialogFragment.Controller {

    public static final String ALERT_DIALOG_TAG = "alert_dialog_tag";
    private static final String TAG_LIST_FRAGMENT = "TAG_LIST_FRAGMENT";
    private static final String TAG = "@@@";
    private int pressedCount = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if (savedInstanceState == null) {
            showListInMainContainer();
        }
    }

    private void showListInMainContainer() {
        Fragment notesFragment = new NotesFragment();
        getSupportFragmentManager()
                .beginTransaction()
                .add(R.id.activity_main__main_fragment_container, notesFragment, TAG_LIST_FRAGMENT)
                .commit();
    }

    @Override
    public void showNoteDetails(Note note) {
        Fragment noteDetailsFragment = NoteDetailsFragment.newInstance(note);
        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.activity_main__second_fragment_container, noteDetailsFragment)
                .addToBackStack(null)
                .commit();
    }

    @Override
    public void showNewNoteDialog() {
        new AlertDialogFragment()
                .show(getSupportFragmentManager(), ALERT_DIALOG_TAG);
    }

    @Override
    public void popBackFragment() {
        getSupportFragmentManager().popBackStack();
    }

    @Override
    public void updateDataSet() {
        NotesFragment notesFragment = (NotesFragment) getSupportFragmentManager().findFragmentByTag(TAG_LIST_FRAGMENT);
        if (notesFragment == null)
            throw new IllegalStateException(getString(R.string.ColorsListFragment_not_on_screen));
        notesFragment.updateDataSet();
        notesFragment.scrollToAdded();
    }

    @Override
    public void onBackPressed() {
        if (getSupportFragmentManager().getBackStackEntryCount() == 0) {
            if (++pressedCount % 2 == 0)
                super.onBackPressed();
            else {
                View view = findViewById(R.id.activity_main__main_fragment_container);
                Snackbar
                        .make(this, view, getString(R.string.backpressed_message), Snackbar.LENGTH_SHORT)
                        .setAction(R.string.snackbar_button_label, v -> super.onBackPressed())
                        .setActionTextColor(ContextCompat.getColor(this, android.R.color.holo_red_light))
                        .show();
            }
        } else {
            pressedCount = 0;
            super.onBackPressed();
        }
    }

    @Override
    public void addNewNote(String title, String content) {
        App.get().notes.addNewNote(title, content);
        updateDataSet();
    }
}

File: app/src/main/java/me/bkkn/ui/MainActivity.kt
package me.bkkn.ui

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import me.bkkn.App
import me.bkkn.databinding.ActivityMainBinding
import me.bkkn.domain.entity.Note
import me.bkkn.domain.repository.Notes
import me.bkkn.ui.list.NoteAdapter
import me.bkkn.ui.list.NoteViewHolder.OnNoteListener

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private var notes: Notes? = null
    private var adapter: NoteAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMainBinding.inflate(layoutInflater)

        setContentView(binding.root)
        notes = App.get(this).notes
        initRecycler()
    }

    private fun initRecycler() {
        binding.addNoteButton.setOnClickListener {
            notes!!.addNewNote()
            val list = notes!!.notes
            adapter!!.setData(list)
            adapter!!.notifyItemInserted(list.size - 1)
            binding.recyclerView.smoothScrollToPosition(list.size - 1)
        }
        binding.recyclerView.layoutManager = LinearLayoutManager(this)
        adapter = NoteAdapter()
        val list = notes!!.notes
        adapter!!.setData(list)
        adapter!!.setOnDeleteClickListener(object : OnNoteListener {
            override fun onDeleteNote(note: Note) {
                notes!!.deleteNote(note)
                adapter!!.setData(notes!!.notes)
            }

            override fun onClickNote(note: Note) {
                val intent = Intent(this@MainActivity, NoteActivity::class.java)
                intent.putExtra(NoteActivity.NOTE_EXTRA_KEY, note)
                startActivityForResult(intent, NOTE_REQUEST_CODE)
            }
        })
        binding.recyclerView.setAdapter(adapter)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == NOTE_REQUEST_CODE && resultCode == RESULT_OK) {
            adapter!!.setData(notes!!.notes)
            //Note note = data.getParcelableExtra(NOTE_EXTRA_KEY);
            //int idx = App.get(this).getNotes().index(note);
            adapter!!.notifyDataSetChanged()
        }
    }

    companion object {
        private const val NOTE_REQUEST_CODE = 42
    }
}